File: src/main/java/com/dingdongdeng/coinautotrading/common/client/util/QueryParamsConverter.java
package com.dingdongdeng.coinautotrading.common.client.util;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.Map;
import java.util.Objects;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;

@RequiredArgsConstructor
@Component
public class QueryParamsConverter { //fixme 적절한 이름으로 수정필요

    private final ObjectMapper objectMapper;

    public MultiValueMap<String, String> convert(Object object) {
        try {
            MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
            if (Objects.isNull(object)) {
                return params;
            }
            Map<String, String> map = objectMapper.convertValue(object, new TypeReference<Map<String, String>>() {
            });
            params.setAll(map);
            return params;
        } catch (Exception e) {
            throw new IllegalStateException("fail generate query params", e);
        }
    }
}

File: app-service/src/main/kotlin/com/dingdongdeng/autotrading/infra/client/common/QueryParamsConverter.kt
package com.dingdongdeng.autotrading.infra.client.common

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.stereotype.Component
import org.springframework.util.LinkedMultiValueMap
import org.springframework.util.MultiValueMap
import org.springframework.web.util.UriComponentsBuilder

@Component
class QueryParamsConverter(
    private val objectMapper: ObjectMapper
) {
    fun convertMap(body: Any): MultiValueMap<String, String> {
        return try {
            val params: MultiValueMap<String, String> = LinkedMultiValueMap()
            val map = objectMapper.convertValue(body, object : TypeReference<Map<String, Any>>() {})

            map.forEach { (key: String, value: Any) ->
                if (value is List<*>) { //fixme 개선필요
                    value.toList().forEach { params.add(key, it.toString()) }
                    return@forEach
                }
                params.add(key, value.toString())
            }
            params
        } catch (e: Exception) {
            throw IllegalStateException("fail generate query params", e)
        }
    }

    fun convertStr(body: Any): String {
        return UriComponentsBuilder.fromUriString("")
            .queryParams(convertMap(body))
            .build() //.encode()
            .toUriString()
    }
}
